smart_open/s3.py

```diff
@@ -6,16 +6,23 @@
 # from the MIT License (MIT).
 #
 """Implements file-like objects for reading and writing from/to S3."""
+import sys
+import warnings
+warnings.warn("Please install via smart_open[gcs] as we will be dropping cloud dependency installs by default", DeprecationWarning)
 
 import io
 import functools
 import logging
 import time
 
-import boto
-import boto3
-import botocore.client
-import botocore.exceptions
+try:
+    import boto
+    import boto3
+    import botocore.client
+    import botocore.exceptions
+except ImportError:
+    sys.stderr.write("Install via smart_open[aws] or smart_open[all] to use this module")
```

Comment:
Why this instead of a logger.info() ? Otherwise the message will be displayed all the time for every library that is not installed

Reply:
or a least the s3 import needs to be taken out from __init__.py otherwise the actual version will sys.exit(1) without boto3 installed

Reply:
The logger will only print if the root logger is configured which is no guarantee.

As far as that import path in __init__.py goes it would seem that it would need to be removed. Once those underlying dependencies are dropped from being installed I believe the import will cause issues. If sys.exit is dropped and the shortcut path maintained it would cause the import error to be suppressed here and thrown elsewhere. Whereas by maintaining the sys.exit the program exits exactly where the message to fix the issue is printed vs exiting with an import error elsewhere that doesn't immediately tell you what the issue is.

Reply:
We would probably want a deprecation warning for the import path in this PR.

Reply:
I was unaware of that import path so thank you for pointing that out.

Reply:
Yeah so didn't see that in `transport.py` sys.exit will cause an issue every time. So yeah this is a bigger issue than just the short import path in `__init__.py`

Reply:
Maybe move `register_transport('smart_open.s3')` to after the `try/except` block in each of the cloud modules?

Reply:
^ moving the register after try/except did not work so I moved it back. Also I modified the try/except block to reraise the `ImportError` as the `register_transport` would never see it w/o it.

Reply:
We have two separate problems.

1. We import s3.iter_bucket in __init__.py. This will explode if boto3 is not installed. Currently, we _always_ install boto3 (for backwards compatibility) so it's not an acute problem.
2. When we try to register all the transports, we log a warning for each submodule we couldn't import. This will be annoying for people who intentionally didn't install dependencies for a particular submodule.

I think the solution for (1) is simple. When we stop installing boto3 by default, we can try-catch around the culprit import. We can also remove the import altogether, and direct people to use `smart_open.s3.iter_bucket` via a warning.

I think the solution for (2) is to group the submodules that failed to register together, and then log a single warning like:

```
The following submodules could not be imported due to missing dependencies: abs, s3, gcs. To use those submodules, you must install their dependencies.  See <URL> for more info.
```

Once people get used to the new extra handling (a few versions later), we can reduce the verbosity of the log message.

@Amertz08 I'd like to include your PR in a soon-to-be bugfix release of smart_open (we need to do one because the Google Cloud deps are messing things up for some people). Can you please roll it back to the state it was in when I approved it? I will then merge. We can continue our discussion and work in a later PR.

Reply:
(Please roll back all commits after 0d86f9f. Don't force push, it will axe our discussions).